cmd/SplitFFISSpreadsheet/sheet.go

```diff
@@ -0,0 +1,210 @@
+package main
+
+import (
+	"io"
+	"net/url"
+	"regexp"
+	"strconv"
+	"time"
+
+	"github.com/usdigitalresponse/grants-ingest/internal/log"
+	"github.com/usdigitalresponse/grants-ingest/pkg/grantsSchemas/ffis"
+	"github.com/xuri/excelize/v2"
+)
+
+// Currently, the FFIS spreadsheet uses an "X" to indicate eligibility
+func parseEligibility(value string) bool {
+	return value == "X"
+}
+
+// parseXLSXFile is a function that reads and processes an Excel file stream, converting the data into a slice
+// of ffis.FFISFundingOpportunity objects. The file is expected to be provided as an io.Reader.
+// The function filters and retains only those funding opportunities that possess a valid grant ID.
+//
+// Any errors encountered during the parsing of individual cells within the Excel file are not returned as function errors,
+// but are instead logged at the WARN level, accompanied by the associated row and column indices for easy identification.
+//
+// Parameters:
+// r: The io.Reader providing the Excel file stream to be parsed.
+// logger: The log.Logger used to log any parsing errors at the WARN level.
+//
+// Returns:
+// A slice of ffis.FFISFundingOpportunity objects representing the parsed funding opportunities from the Excel file.
+// An error is returned if the parsing process fails at a level beyond individual cell parsing.
+func parseXLSXFile(r io.Reader, logger log.Logger) ([]ffis.FFISFundingOpportunity, error) {
+	xlFile, err := excelize.OpenReader(r)
+
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		if err := xlFile.Close(); err != nil {
+			log.Error(logger, "Error closing excel file", err)
+		}
+	}()
+
+	// We assume the excel file only has one sheet
+	sheet := "Sheet1"
+
+	// Returns all rows in the sheet. Note this assumes the sheet is somewhat limited in
+	// size, and will not scale to extremely large worksheets (memory overhead)
+	rows, err := xlFile.GetRows(sheet)
+	if err != nil {
+		return nil, err
+	}
+
+	var opportunities []ffis.FFISFundingOpportunity
+
+	// Tracks if the iterator has found headers for the sheet. A header
+	// is a column header, like "CFDA", "Opportunity Title", etc.
+	foundHeaders := false
+
+	// The sheet has rows as bills that we want to assign to
+	// each opportunity underneath that bill, so we use this
+	// as we iterate through the rows. This could be something like
+	// "Inflation Reduction Act".
+	bill := ""
+
+rowLoop:
+	for rowIndex, row := range rows {
+		opportunity := ffis.FFISFundingOpportunity{}
```

Comment:
I ended up doing the row_count metric, which is a great idea – I feel the `rows_read` one is less useful as it will log no matter what n times (where n is row count). I did like the idea of monitoring the trend of spreadsheet rows / parsed opportunities, so I added a metric to record the parsed opportunities in the handler.